File: src/Test/PolymorphismFFI.rs
struct Monoidish<A> {
    mempty_: A,
    mappend_: Box<dyn Fn(A) -> Box<dyn Fn(A) -> A>>,
}

fn poly_loop<A: Clone>(dict: &Monoidish<A>, mut limit: i64, mut acc: A) -> A {
    while limit > 0 {
        let f = (dict.mappend_)(acc);
        acc = f(dict.mempty_.clone());
        limit -= 1;
    }
    acc
}

pub fn Test_PolymorphismFFI_runPolymorphismFFI(limit: i64) -> i64 {
    let int_monoidish = Monoidish {
        mempty_: 1,
        mappend_: Box::new(|x| Box::new(move |y| x + y)),
    };
    poly_loop(&int_monoidish, limit, 0)
}

File: src/Test/PolymorphismFFI.cc
#include "purescript.h"
#include <functional>

namespace {
    template <class A> using Endo = std::function<A(A)>;
    template <class A>
    struct Monoidish {
        A mempty;
        std::function<Endo<A>(A)> mappend;
    };

    template <class A>
    A polyLoop(const Monoidish<A>& dict, int n, A acc) {
        while (n > 0) {
            acc = dict.mappend(acc)(dict.mempty);
            --n;
        }
        return acc;
    }
}

FOREIGN_BEGIN(Test_PolymorphismFFI)
exports["runPolymorphismFFI"] = [](const boxed& in) -> boxed {
    const Monoidish<int> dict{1, [](int x) -> Endo<int> {
        return [x](int y) { return x + y; };
    }};
    const int result = polyLoop(dict, unbox<int>(in), 0);
    return result;
};
FOREIGN_END
